### PR TITLE
Fixes ARP messages from being sent through wrong endpoints

### LINKED ISSUE #1088: [BUG] ARP sends packets over the wrong network interfaces.
**Describe the bug**
When multiple interfaces and endpoints are present, ARP leaks out one NetIf's IPv4 addresses over the other NetIfs.

**Target**
- ALL

**To Reproduce**
ARP_Leakage
Check out "before.pcapng" in the attached zip file The DUT send out Gratuitous ARP's which in the capture look like ARP requests. the protocol addresses for these ARP requests are 10.238.238.10 and 192.168.21.47. Those last two addresses belong to the Fake0 interface and should not finding their way to different interfaces. 

**Expected behavior**
After a quick fix, "after.pcapng" from the .zip file shows the resulting ARP packets. I'm pretty sure this should be the correct behavior. where we only see ARP packets associated with 172.30.30.26

**Wireshark logs**
[ARP_Leakage.zip](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/files/14140394/ARP_Leakage.zip)

I have a fix for this and was planning on starting a PR. Before I do that though, I'd like to have a discussion on the issue. Here are my thoughts and questions:

1. Do you agree that this is a bug. I strongly believe it is. I have never seen a host do this and I can't come up with even a theoretical reason for this behavior... but I have been wrong many times before, so please let me know if you think I'm missing something.
2. The offending code is `FreeRTOS_OutputARPRequest()` which sends out an ARP request on all available IPv4 endpoints.  My first quick and dirty fix was to use `FreeRTOS_FindEndPointOnNetMask()` inside `FreeRTOS_OutputARPRequest()` and only send to the found end-point if there was one. This approach removes a for(;;) loop and one if() clause from `FreeRTOS_FindEndPointOnNetMask()` making it substantially cleaner, but I think there might be an even better fix...
3. Another solution ( and I think that is the better one ) would be to add a `NetworkEndPoint_t * pxEndPoint` parameter to `FreeRTOS_OutputARPRequest()` I need to look deeper into this solution, but here's my reasoning: `FreeRTOS_OutputARPRequest()` has one very specific function and almost every function that calls it already know which endpoint must be used. I believe `FreeRTOS_OutputARPRequest()`  should not be burdened with searching for end-points. I have looked at all calls to `FreeRTOS_OutputARPRequest()` and all of them except two already know the endpoint that should be used so they can simply give it to `FreeRTOS_OutputARPRequest()`. 
4. There are 2 exceptions to the above:
5. `xARPWaitResolution()` calls `FreeRTOS_OutputARPRequest()`  without knowing the proper endpoint, BUT..... `xARPWaitResolution()` is never used so it's optimized out, AND even if it were used, I believe it needs to use `FreeRTOS_FindEndPointOnNetMask()` and absorb the burden of the search before calling `FreeRTOS_OutputARPRequest()` with the proper endpoint already figured out.
6. The second exception to (3) is `prvTCPPrepareConnect_IPV4()` As is the case with (5) I believe `prvTCPPrepareConnect_IPV4()` should  find the end-point before calling `FreeRTOS_OutputARPRequest()`
7. I am willing to create a PR for the more elaborate solution listed in (3)
8. I can even have both solutions as draft PRs and we can compare them before choosing what to do

Let me know what you think and thank you for your time.

### PR DESCRIPTION
**Pull request by tony-josi-aws**
Description
-----------
This PR fixes #1088, by adding a new API `FreeRTOS_OutputARPRequest_Multi` which requires the interface to be targeted while sending the ARP messages. 

This PR also fixes some variable declarations that are present after code.

Test Steps
-----------

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

#1088 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/FreeRTOS_ARP.c [hunk 1]
```diff
@@ -1338,6 +1340,24 @@ static BaseType_t prvFindCacheEntry( const MACAddress_t * pxMACAddress,
...
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Create and send an ARP request packet.
+ *
+ * @param[in] ulIPAddress A 32-bit representation of the IP-address whose
+ *                         physical (MAC) address is required.
+ */
+    void FreeRTOS_OutputARPRequest( uint32_t ulIPAddress )
```
Comment (moninom1) [inline source/FreeRTOS_ARP.c:1352]: It should enclosed with ipconfigIPv4_BACKWARD_COMPATIBLE flag.

File: source/FreeRTOS_DHCP.c [hunk 2]
```diff
@@ -1455,7 +1447,7 @@
                  * it easier to lookup a device in a router's list of DHCP clients. */
 
                 /* Point to where the OPTION_END was stored to add data. */
-                pucPtr = &( pucUDPPayloadBuffer[ dhcpFIRST_OPTION_BYTE_OFFSET + ( *pxOptionsArraySize - 1U ) ] );
+                uint8_t * pucPtr = &( pucUDPPayloadBuffer[ dhcpFIRST_OPTION_BYTE_OFFSET + ( *pxOptionsArraySize - 1U ) ] );
```
Comment (moninom1) [inline source/FreeRTOS_DHCP.c:1450]: uint8_t * pucPtr should be declared at line 1400, any reason for moving it here?

File: source/FreeRTOS_TCP_IP_IPv6.c [hunk 3]
```diff
@@ -100,6 +100,15 @@ BaseType_t xProcessReceivedTCPPacket_IPV6( NetworkBufferDescriptor_t * pxDescrip
 {
     /* Function might modify the parameter. */
     NetworkBufferDescriptor_t * pxNetworkBuffer = pxDescriptor;
+    const ProtocolHeaders_t * pxProtocolHeaders;
```
Comment (moninom1) [inline source/FreeRTOS_TCP_IP_IPv6.c:103]: Good cleanup, was it coming as a warning or MISRA error?

File: source/FreeRTOS_ARP.c (hunk 1 above)
Comment (tony-josi-aws) [reply source/FreeRTOS_ARP.c:1352]: Hey Monika, `FreeRTOS_OutputARPRequest` is used in the stack when there isn't an endpoint readily available and have to use `FreeRTOS_FindEndPointOnNetMask` (basically what `FreeRTOS_OutputARPRequest` does). Hence, it is not a function kept for backward compatibility.

File: source/FreeRTOS_DHCP.c (hunk 2 above)
Comment (tony-josi-aws) [reply source/FreeRTOS_DHCP.c:1450]: `pucPtr` is only used when `#if ( ipconfigDHCP_REGISTER_HOSTNAME == 1 )` evaluates to true.

File: source/FreeRTOS_TCP_IP_IPv6.c (hunk 3 above)
Comment (tony-josi-aws) [reply source/FreeRTOS_TCP_IP_IPv6.c:103]: TY, its part of the C90 compliance changes not MISRA.

File: source/FreeRTOS_ARP.c (hunk 1 above)
Comment (moninom1) [reply source/FreeRTOS_ARP.c:1352]: Make sense.

**Review by moninom1 — DISMISSED** [review]

**Review by ActoryOu — COMMENTED** [review]
Could you help add description on why some `FreeRTOS_OutputARPRequest` are changed but some are still same?

File: test/cbmc/proofs/IP/ProcessEthernetPacket/ProcessEthernetPacket_harness.c [hunk 4]
```diff
@@ -87,5 +89,8 @@ void harness()
     pxNetworkBuffer->pucEthernetBuffer = ( ( ( uint8_t * ) safeMalloc( pxNetworkBuffer->xDataLength ) ) );
     __CPROVER_assume( pxNetworkBuffer->pucEthernetBuffer != NULL );
 
+    /* Its assumed that every RX packet will have a valid network interface assigned. (Asserted in code) */
+    pxNetworkBuffer->pxInterface = &xInterface;
```
[the remark this answers is not in the document]
Comment (tony-josi-aws) [reply test/cbmc/proofs/IP/ProcessEthernetPacket/ProcessEthernetPacket_harness.c:93]: In #1096 we have [made explicit checks](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/1096/files#diff-8b61e00e54b55092bc30ba18e0dc9570da3218e0b816bd15d47ba3b81c054318R1556) to make sure that the interface or endpoint of an RX packet is never NULL.

File: source/FreeRTOS_ARP.c [hunk 5]
```diff
@@ -1338,6 +1340,24 @@ static BaseType_t prvFindCacheEntry( const MACAddress_t * pxMACAddress,
...
+    {
+        const NetworkEndPoint_t * pxEndPoint = FreeRTOS_FindEndPointOnNetMask( ulIPAddress, 12 );
```
Comment (shubnil) [inline source/FreeRTOS_ARP.c:1354]: Please add a comment mentioning the below assumptions :
1. IPv4 endpoints belonging to same physical interface can have same subnet
2. IPv4 endpoints belonging to different physical interface in the system will have a different subnet.

**Review by shubnil — COMMENTED** [review]
The code looks good. We can add the assumptions for subnet in both the code as well in the endpoint documentation.

**Review by ActoryOu — DISMISSED** [review]

**Review by moninom1 — DISMISSED** [review]

**Review by moninom1 — APPROVED** [review]

**Review by moninom1 — DISMISSED** [review]

**Review by shubnil — DISMISSED** [review]

**Comment by htibosch** [comment]
The addition of the function `FreeRTOS_OutputARPRequest_Multi()` is very useful.

However, at the moment the function is called, we already have a reliable endpoint pointer. So I think there is no need to iterate through all endpoints.

In your PR, I see the following loop:

~~~c
void FreeRTOS_OutputARPRequest_Multi( const NetworkInterface_t * pxNetworkInterface,
                                      uint32_t ulIPAddress )
{
    /* Iterate through all endpoints that are attached to 'pxNetworkInterface'. */
    for( pxEndPoint = FreeRTOS_FirstEndPoint( pxNetworkInterface );
         pxEndPoint != NULL;
         pxEndPoint = FreeRTOS_NextEndPoint( pxNetworkInterface, pxEndPoint ) )
    {
        if( ( pxEndPoint->bits.bIPv6 == pdFALSE_UNSIGNED ) &&
            ( pxEndPoint->ipv4_settings.ulIPAddress != 0U ) )
        {
            // As for the endpoint, you are using the first IPv4 endpoint which has an IP-address
            break;
        }
    }
}
~~~

I am using the following configuration for a test:
~~~c
    SetupV4_t setup_v4_first =
    {
        .pcMACAddress = "00-11-11-11-11-07",
        .pcIPAddress = "172.16.0.104",
        .pcMask = "255.255.224.0",
        .eType = eStatic,  /* or eDHCP if you wish. */
    };
    SetupV4_t setup_v4_second =
    {
        .pcMACAddress = "00-11-11-11-11-02",
        .pcIPAddress = "192.168.2.104",
        .pcMask = "255.255.255.0",
        .pcGateway = "192.168.2.1",
        .pcDNS[0] = "118.98.44.10",
        .eType = eStatic,
    };
    xSetupEndpoint_v4( &xInterfaces[ 0 ], &( xEndPoints[ 0 ] ), &setup_v4_first );
    xSetupEndpoint_v4( &xInterfaces[ 0 ], &( xEndPoints[ 1 ] ), &setup_v4_second );
~~~
( using [net_setup.[ch]](https://github.com/htibosch/freertos_plus_projects/blob/master/plus/Framework/FreeRTOS-Plus-TCP-main/tools/tcp_utilities/net_setup.c) to configure the endpoints. )

When querying gateway "192.168.2.1" with ARP,  queries are still sent using wrong endpoints. The above function selects the first IPv4 endpoint, which is on the wrong network "172.x.x.x", in stead of "192.168.2.x".
When I swap the order of the declarations, the problem does not occur.

Your test-case: when I use two separate interfaces, the problem does not occur neither. But when I attach the 2 endpoints to the same interface, ARP does not work properly.

So my proposal is the following:
~~~c
void FreeRTOS_OutputARPRequest_Multi( const NetworkEndPoint_t * pxEndpoint,
                                      uint32_t ulIPAddress )
{
    /* This is called from the context of the IP event task, so a block time
     * must not be used. */
    pxNetworkBuffer = pxGetNetworkBufferWithDescriptor( sizeof( ARPPacket_t ), ( TickType_t ) 0U );
    // etc
}
~~~
You will have to change the calls in 4 locations, which is straightforward.

After the above change, the order of declaration of the endpoints doesn't matter anymore.

Regards

**Comment by evpopov** [comment]
Hein,
I think you are referring to #1105
<https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/1105> and not my PR. My
PR was #1104 <https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/1104>
My PR did not add a new function, It just fixed the
old FreeRTOS_OutputARPRequest() function.
After Tony opened #1105, I closed my PR.
I suggested to Tony that he include all the changes from my PR because
after the call to FreeRTOS_FindEndPointOnNetMask() we still need to make
sure that ulIPAddress is != 0 and != ~0 before we send an ARP request. They
haven't included tose checks in #1105 yet, but I hope they do eventually.

**Comment by htibosch** [comment]
@evpopov wrote:
> Hein, I think you are referring to #1105 and not my PR. My PR was #1104

Yes indeed, I was testing Tony's PR about `FreeRTOS_OutputARPRequest()`.
My remarks also apply to your #1104.

**Comment by tony-josi-aws** [comment]
Hi @htibosch,

>         if( ( pxEndPoint->bits.bIPv6 == pdFALSE_UNSIGNED ) &&
>            ( pxEndPoint->ipv4_settings.ulIPAddress != 0U ) )
>        {
>           // As for the endpoint, you are using the first IPv4 endpoint which has an IP-address
>            break;
>       }

The `FreeRTOS_OutputARPRequest_Multi` API doesn't have a `break`, the requests are sent through all the endpoints of the given interface.

Though the ARP request to the first endpoint is not necessary in this case, it will be sent through the second endpoint as well.

**Comment by htibosch** [comment]
Hello @tony-josi-aws,
> The `FreeRTOS_OutputARPRequest_Multi` API doesn't have a `break`, the requests are sent through all the endpoints of the given interface.
> 
> Though the ARP request to the first endpoint is not necessary in this case, it will be sent through the second endpoint as well.

I haven't had the time to check this thought:

**Without** the break, the ARP packet will be sent to each IPv4 endpoint that it up, but with a different return address. My router on 192.168.2.1 will reply to it, but the there is a problem: the ARP cache entry already has an entry for 192.168.2.1, but with a different endpoint.

**With** the break, there is 50% chance that it works for me, that depends on the order in which de endpoints were added.

File: source/FreeRTOS_ARP.c [hunk 6]
```diff
@@ -1264,77 +1264,71 @@ static BaseType_t prvFindCacheEntry( const MACAddress_t * pxMACAddress,
...
  *                         physical (MAC) address is required.
  */
-    void FreeRTOS_OutputARPRequest_Multi( const NetworkInterface_t * pxNetworkInterface,
+    void FreeRTOS_OutputARPRequest_Multi( const NetworkEndPoint_t * pxEndPoint,
                                           uint32_t ulIPAddress )
     {
         NetworkBufferDescriptor_t * pxNetworkBuffer;
-        NetworkEndPoint_t * pxEndPoint;
 
         /* Send an ARP request to every end-point which has the type IPv4,
```
Comment (moninom1) [inline source/FreeRTOS_ARP.c:1278]: Please remove the comment.

**Review by htibosch — APPROVED** [review]
Thank you Tony so far, I approve the changes

File: test/cbmc/proofs/ARP/ARPProcessPacket/ARPProcessPacket_harness.c [hunk 7]
```diff
@@ -8,12 +8,14 @@
...
 #include "FreeRTOS_IP_Private.h"
 #include "FreeRTOS_ARP.h"
+#include "FreeRTOS_Routing.h"
 
 /* This pointer is maintained by the IP-task. Defined in FreeRTOS_IP.c */
 extern NetworkBufferDescriptor_t * pxARPWaitingNetworkBuffer;
 
 /* This is an output function and need not be tested with this proof. */
-void FreeRTOS_OutputARPRequest( uint32_t ulIPAddress )
+void FreeRTOS_OutputARPRequest_Multi( NetworkEndPoint_t * pxEndPoint,
```
Comment (moninom1) [inline test/cbmc/proofs/ARP/ARPProcessPacket/ARPProcessPacket_harness.c:17]: As we are removing FreeRTOS_OutputARPRequest, don't we have the case where flow is from FreeRTOS_OutputARPRequest to -> FreeRTOS_OutputARPRequest_Multi?

**Review by moninom1 — APPROVED** [review]

**Review by shubnil — APPROVED** [review]

**Pull request merged**